src/MongoDB.Driver.Core/Core/Bindings/ReadPreferenceBinding.cs

```diff
@@ -68,15 +69,18 @@ public ICoreSessionHandle Session
         public IChannelSourceHandle GetReadChannelSource(CancellationToken cancellationToken)
         {
             ThrowIfDisposed();
-            var server = _cluster.SelectServer(_serverSelector, cancellationToken);
+            var server = BindingHelper.SelectServer(_session, _cluster, _serverSelector, cancellationToken);
+            BindingHelper.PinShardRouterToSessionIfNeeded(_session, _cluster, server);
```

Comment:
I think you can move the call to PinShardRouterToSessionIf to inside the SelectServer method, since every call to SelectServer MUST currently be followed by a call to PinShardRouterToSessionIfNeeded.

Reply:
Hm, that save an extra LOC per GetChannelSource, true. Instead of moving `PinShardRouterToSession` inside SelectServer, how about creating a new method `SelectAndPinServerIfNeeded`  that calls the existing `SelectServer` and `PinShardRouterToSessionIfNeeded`. That way, `SelectServer`'s responsibility is purely  selecting a server, and `PinShardRouterToSessionIfNeeded`'s responsibility is purely whether or not we need to mutate the session by pinning the `mongos`. Meanwhile, the Binding classes still get to save a line of code, but it's a little more obvious that the helper function they're calling is doing a bit more than simply selecting and returning a server?

Reply:
It's not about saving a line of code.
It's about the caller maybe forgetting to call the second method.
That's why it's better to have a single method that does both.
Not sure it's worth breaking into two methods internally, but if you do then the two helper methods should be private so that the caller can only call the public method that does both.

Reply:
Ah okay, thanks for clarifying my misunderstanding :) I think I'd prefer to have two methods internally: I'll keep them private and only expose one public method per your suggestion!